libs/autotagging-commons/src/main/java/org/opensearch/rule/service/IndexStoredRulePersistenceService.java

```diff
@@ -0,0 +1,266 @@
+        ClusterService clusterService,
+        Client client,
+        FeatureType featureType,
+        int maxRulesPerGetRequest
+    ) {
+        this.indexName = indexName;
+        this.clusterService = clusterService;
+        this.client = client;
+        this.featureType = featureType;
+        this.maxRulesPerGetRequest = maxRulesPerGetRequest;
+    }
+
+    /**
+     * Entry point for the create rule api logic in persistence service
+     * @param request - The CreateRuleRequest
+     * @param listener - ActionListener for CreateRuleResponse
+     */
+    public void createRule(CreateRuleRequest request, ActionListener<CreateRuleResponse> listener) {
+        try (ThreadContext.StoredContext ctx = getContext()) {
+            createIndexIfAbsent(new ActionListener<>() {
+                @Override
+                public void onResponse(Boolean indexCreated) {
+                    if (!indexCreated) {
+                        listener.onFailure(new IllegalStateException(indexName + " index creation failed and rule cannot be persisted"));
+                        return;
+                    }
+                    checkDuplicateRule(request.getRule(), listener);
+                }
+
+                @Override
+                public void onFailure(Exception e) {
+                    listener.onFailure(e);
+                }
+            });
+        }
+    }
+
+    /**
+     * Check if there's an existing Rule with the same attributes.
+     * For example, if there's an existing Rule with the attribute index_pattern: ["a", "b", "c"],
+     * then we cannot create another Rule with only one attribute index_pattern: ["b"], because the value "b"
+     * already exists under another Rule. Note that the conflict exists only when we have the exact same attribute
+     * names in the two rules (That is, a Rule with attribute "index_pattern" won't create a conflict with another
+     * Rule that has "index_pattern" and some other attributes).
+     * @param rule - The rule to update.
+     * @param listener - ActionListener for CreateRuleResponse
+     */
+    public void checkDuplicateRule(Rule rule, ActionListener<CreateRuleResponse> listener) {
+        try (ThreadContext.StoredContext ctx = getContext()) {
+            getRuleFromIndex(null, rule.getAttributeMap(), null, new ActionListener<>() {
+                @Override
+                public void onResponse(GetRuleResponse getRuleResponse) {
+                    String duplicateRuleId = IndexStoredRuleUtils.getDuplicateRuleId(rule.getAttributeMap(), getRuleResponse.getRules());
+                    if (duplicateRuleId != null) {
+                        listener.onFailure(
+                            new IllegalArgumentException(
+                                "A rule that has the same attribute values already exists under rule id " + duplicateRuleId
+                            )
+                        );
+                        return;
+                    }
```

Comment:
I think we can improve this using `Optional<String>` and use `result.map(throwException).orElse(persistRule)`